### ISSUE TITLE
Way to override method in shared library?

### ISSUE THREAD

**Issue by idij** [lesfurets/JenkinsPipelineUnit#50]
I'm unit testing a shared library, which I load in the setUp into a variable `library`.

each (spock) test of the shared library then takes the form:
``` groovy
given: "a mock bar"
    def bar = Mock (Closure)
    helper.registerAllowedMethod ('bar', [String.class], bar)
when: "foo is called"
    library.foo()
then: "bar is called"
    1 * bar (_)
```

This works fine if `bar` is some jenkins method (e.g `sh` or something), but if `bar` is a different method in the same shared library then registration doesn't replace that `bar` with the mock for the test. 

This means that I can only test the library itself as a unit (`foo` + `bar` + mock of bar's externals) rather than each method as a self contained unit (`foo` + mock of foo's externals including `bar`). 

If `bar` is complex, then the mocking to exercise `foo` becomes correspondingly complex, when really all I want to confirm in the is that `foo` calls `bar`.

Is there a way to achieve the replacement of an arbitrary library method in this way? I know I can probably "load" the foo.groovy directly, but that requires replacing all of the potentially complex library dependencies that foo has rather than just the interesting bits.

**Comment by headcrabmeat** [comment]
Hi, I hope you can find it useful https://github.com/lesfurets/JenkinsPipelineUnit/issues/49

Unfortunately I'm not sure how to align it with Spoke, but what i'm using is  
```
def le_mock = new MockFor(SomeClass);
le_mock.demand.someMethod{println "I've been mocked"};

le_mock.use {
   // script to load my shared lib
    def script = loadScript("template/pipeline/template.groovy")
}
```
It works fine for jenkins internals (e.g. jenkins.model.Jenkins), however if I try to mock a class in my shared lib it is really not in the groovyClassPath. Actually I wasn't succeed to configure class path with gradle. So I cannot prove it would work.  

However this framework is about how to use `groovyClassLoader`. So I think we can try to find the `Class` object of the class we want to mock using 
`this.class.classLoader` and then mock it it would work

**Comment by idij** [comment]
I figured it out,

for the jenkins methods, since they don't exist, the registered signature can be anything so long as it matches the arguments, and generally I'd been using  `[Map.class]`, but for the shared library methods it has to match the exact signature of the declaration and it's not sufficient to just match the calling params. So if you have:

**vars/bar.groovy:**
``` groovy
def call (args) {
    if (args.baz) { /* ... */ }
}
```

Even though args is being used as a `Map`, and if you asked people what is args, they'd probably say "it's a map", the actual signature is of an `Object`. As a result, `[Map.class]` even though it would match what you call it with, and is fine for a similar Jenkins function, doesn't match the declared signature of the shared library function. I had to either register it as `[Object.class]` or make args explicitly a `Map`.
